### PR TITLE
Streams: reject pending reads when releasing reader

### PR DESCRIPTION
**Pull request by MattiasBuelens**
See https://github.com/whatwg/streams/pull/1168 for the accompanying spec change.

### CODE REVIEW

File: streams/readable-byte-streams/bad-buffers-and-views.any.js [hunk 1]
```diff
@@ -203,6 +203,38 @@ async_test(t => {
...
+
+      assert_throws_js(RangeError, () => c.byobRequest.respondWithNewView(view));
+    }),
+    type: 'bytes'
+  });
+  const reader = stream.getReader({ mode: 'byob' });
+
+  reader.read(new Uint8Array([4, 5, 6]));
+}, 'ReadableStream with byte source: respondWithNewView() throws if the supplied view has a different offset ' +
+   '(in the readable state)');
```
Comment (MattiasBuelens) [inline streams/readable-byte-streams/bad-buffers-and-views.any.js:219]: Drive-by fix: I noticed that [this check](https://github.com/whatwg/streams/blob/53282dde15be5f844768c2167405660da3c29b2d/reference-implementation/lib/abstract-ops/readable-streams.js#L1657-L1659) was not covered by any tests, so I added one. 🙂

File: streams/readable-byte-streams/general.any.js [hunk 2]
```diff
@@ -2393,3 +2404,475 @@ promise_test(async t => {
...
+  assert_not_equals(byobRequest1, null, 'first byobRequest should exist');
+  assert_typed_array_equals(byobRequest1.view, new Uint8Array([1, 2, 3]), 'first byobRequest.view');
+
+  // releaseLock() should reject the pending read, but *not* invalidate the BYOB request
+  reader1.releaseLock();
+  const reader2 = rs.getReader({ mode: 'byob' });
+  const read2 = reader2.read(new Uint8Array([4, 5]));
+  assert_not_equals(controller.byobRequest, null, 'byobRequest should not be invalidated after releaseLock()');
+  assert_equals(controller.byobRequest, byobRequest1, 'byobRequest should be unchanged');
+  assert_array_equals([...new Uint8Array(byobRequest1.view.buffer)], [1, 2, 3], 'byobRequest.view.buffer should be unchanged');
```
Comment (ricea) [inline streams/readable-byte-streams/general.any.js:2515]: I think it's not necessary to unwrap the Uint8Array with `[...]`. assert_array_equals doesn't check the types of the arguments are equal.

File: streams/readable-byte-streams/general.any.js [hunk 3]
```diff
@@ -2393,3 +2404,475 @@ promise_test(async t => {
...
+  // respond(3) should fulfill the second read(), and put 1 remaining byte in the queue
+  byobRequest1.view[0] = 6;
+  byobRequest1.view[1] = 7;
+  byobRequest1.view[2] = 8;
+  byobRequest1.respond(3);
+  const byobRequest2 = controller.byobRequest;
+  assert_equals(byobRequest2, null, 'byobRequest should be null after respond()');
+
+  const result2 = await read2;
+  assert_equals(result2.done, false, 'second result.done');
```
Comment (ricea) [inline streams/readable-byte-streams/general.any.js:2527]: Nit: you can use `assert_false`.

File: streams/readable-byte-streams/general.any.js [hunk 4]
```diff
@@ -2393,3 +2404,475 @@ promise_test(async t => {
...
+  const read2 = reader2.read(new Uint8Array([4, 5, 6]));
+  assert_not_equals(controller.byobRequest, null, 'byobRequest should not be invalidated after releaseLock()');
+  await promise_rejects_js(t, TypeError, read1, 'pending read must reject after releaseLock()');
+
+  // respondWithNewView() should fulfill the *second* read() request
+  byobRequest1.view[0] = 11;
+  byobRequest1.view[1] = 12;
+  byobRequest1.respondWithNewView(byobRequest1.view.subarray(0, 2));
+  const byobRequest2 = controller.byobRequest;
+  assert_equals(byobRequest2, null, 'byobRequest should be null after enqueue()');
```
Comment (ricea) [inline streams/readable-byte-streams/general.any.js:2566]: Should this be `after respondWithNewView()` ?

**Review by ricea — APPROVED** [review]
lgtm

File: streams/readable-byte-streams/general.any.js (hunk 2 above)
Comment (MattiasBuelens) [reply streams/readable-byte-streams/general.any.js:2515]: It matters for [`format_value()`](https://github.com/web-platform-tests/wpt/blob/7915bc605347950da0a0af3707a788048a08ff3c/resources/testharness.js#L1290-L1300) though, when the array ends up in an error message. 🤔

Comment (ricea) [reply streams/readable-byte-streams/general.any.js:2515]: Ah, interesting.
